### ISSUE TITLE
Bug(s) in cv::fitEllipse()

### ISSUE THREAD

**Issue by CaswellBerry** [opencv/opencv#11088]
##### System information (version)
- OpenCV => 3.4
- Operating System / Platform => Windows 10 64bit
- Compiler => Visual Studio 2015

#### This issue affects _**cv::fitEllipse**_ in _**shapedescr.cpp**_.

1. The angle of ellipses with a small difference in width and height (circle-like) are sometimes not assigned to the output-cv::RotatedRect.
2. Furthermore there is an incorrect approximation.
3. Some other lines appear odd to me.

##### Steps to reproduce
First of all: I am no professional and this is my first bug report. Sorry for poor formatting etc.
1. Jumping to the last lines of the function definition: The assignment of the angle value is hidden behind an if-statement which is sometimes not true. Due to this, the default value of 0 is returned with the cv::RotatedRect. The following code snippet shows a simple solution to that.

```
		box.angle = (float)(rp[4] * 180 / CV_PI);	// Assignment of angle moved out of if-bracket

		if (box.size.width > box.size.height)
		{
			float tmp;
			CV_SWAP(box.size.width, box.size.height, tmp);
			box.angle += 90.0;		//  In case w and h swap - 90° should be added
			//box.angle = (float)(90 + rp[4] * 180 / CV_PI);  
		}
```

2. The incorrect approximation is found in line 380.
`	if( fabs(gfp[2]) > min_eps )`
It is not sufficient to check only `gfp[2]`. Instead it should be the whole fraction of `gfp[2] / gfp[1] - gfp[0]`. 

![grafik](https://user-images.githubusercontent.com/37413545/37511606-bfe626ee-28ff-11e8-8a8c-6a4fdb2778f6.png)

Where `g0 = gfp[0], r4 = rp[4]` and so on. The corrected line should look something like this:
`if (fabs(gfp[2]) / fabs(gfp[1] - gfp[0]) > min_eps)`

3. I do not completely understand the applied algorithm in this method but there seem to be two other odd approximations in lines 385 and 388
`    if( rp[2] > min_eps )`
`    if( rp[3] > min_eps )`
If you turn them around, it basically says: If width `(2*rp[2])` and height `(2*rp[3])` are small enough, they can be approximated by the square of their reciprocal value (times 4). In mathematical expression it's something like this:

![grafik](https://user-images.githubusercontent.com/37413545/37511151-0e32eeec-28fe-11e8-8516-459725afbe39.png)

Where `g0 := gfp[0]` and `g1 := gfp[1]`. With `min_eps = 1-e8` as it is hard coded in the function, this cannot be true for any value.

I created a short example project in VS2015 on Windows.

[main.txt](https://github.com/opencv/opencv/files/1818566/main.txt)

**Comment by CaswellBerry** [comment]
I just looked at cv::fitEllipseDirect() and problem (1.) is avoided like that:
```
        if( box.size.width > box.size.height )
        {
            float tmp;
            CV_SWAP( box.size.width, box.size.height, tmp );
            box.angle = (float)(fmod((90 + theta*180/CV_PI),180.0)) ;
        } else {
            box.angle = (float)(fmod(theta*180/CV_PI,180.0));
        };
```

**Comment by saskatchewancatch** [comment]
I agree with @CaswellBerry on issue 1. @CaswellBerry please feel free to fix these in a PR.

**Comment by alalek** [comment]
Related problems in `fitEllipse()` tests:
- `width` / `height` mess in `Size(height, width)` (PR #3936 by @dhood):
https://github.com/opencv/opencv/blob/3.4.2/modules/imgproc/test/test_convhull.cpp#L1446-L1447
- strange 90 angle adjustment of calculated result before check:
    - https://github.com/opencv/opencv/blob/3.4.2/modules/imgproc/test/test_convhull.cpp#L1314-L1318
    - added by this [commit](https://github.com/opencv/opencv_attic/commit/fedece1581bc4f8a77c55b1aa6b263ebf067ac35#diff-1e8e2e547b09e61f7359d1c269daeb4bR1054) from opencv_attic by @vpisarev
- unknown/unspecified normalized (canonical) form of expected `cv::RotatedRect`:
    - `width <= height` (there are many conditions for that) and angle range:  `[0; 180)`
    - or angle range `[0; 90)` and any positive `width` / `height`
    - current implementation is none of these above (negative angles are allowed)

**Comment by ganesh-k13** [comment]
Hey @alalek, I would like to look into this. So after fixing the formula is there anything I can do here?

**Comment by RyanDelGizzi** [comment]
I had issues with fitEllipse returning blatantly incorrect results in a codebase I was working on two years ago.  I caught the issue by rejecting solutions when `angle ==0`. I am revisiting that code now, and wanted to look into the possible cause.  Item 1 here seems to be the likely culprit. 

At the time, I noted this stack overflow [post](https://stackoverflow.com/questions/17159964/opencvs-fitellipse-sometimes-returns-completely-wrong-ellipses) which seems like it may be a victim of the same issue.  

Thanks for looking into this @ganesh-k13, let me know if I can do anything to help.

**Comment by RyanDelGizzi** [comment]
Would this be a good time to document (and test) the output into RotatedRect?

I think that the semi-major and semi-minor axis should always correspond to the width and/or height of the RotatedRect (either way is fine, as long as it is consistent). The angle of the RotatedRect should be from some known axis to the width or height axis.

My intuition says that the semi-major axis should be represented by width, and semi-minor by height. The angle should be the rotation should be from the x-axis to the semi-major axis, bounded either by (-90, 90) or (0,180).

**Comment by rk-exxec** [comment]
Does this problem also affect fitEllipseAMS/Direct and/or OpenCV version 4.x.x?
I am having issues that could be caused by this.
